File: 3D/src/custom_models/Keyframe.java
package custom_models;

import java.io.Serializable;
import java.util.HashMap;

public class Keyframe implements Serializable {
	
	private static final long serialVersionUID = 1L;
	public HashMap<Integer, KeyTransformation> transformations = new HashMap<Integer, KeyTransformation>();
	public int time = 0;
	public double speed = 1.0f;
	public Keyframe(int time) {
		this.time = time;
	}
	
	public Keyframe copy() {
		Keyframe k = new Keyframe(time);
		k.speed = speed;
		for (Integer part : transformations.keySet()) {
			KeyTransformation kt = transformations.get(part);
			k.transformations.put(part, kt.copy());
		}
		return k;
	}

	public void paste(Keyframe k) {
		speed = k.speed;
		transformations.clear();
		for (Integer part : k.transformations.keySet()) {
			KeyTransformation kt = k.transformations.get(part);
			transformations.put(part, kt.copy());
		}
	}
}

File: MonoGame/source/Graphics/Models/Keyframe.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Inignoto.Graphics.Models
{
    public class Keyframe
    {
        public Dictionary<string, KeyTransformation> transformations = new Dictionary<string, KeyTransformation>();
        public int time = 0;
        public double speed = 1.0f;
        public Keyframe(int time)
        {
            this.time = time;
        }

        public Keyframe Copy()
        {
            Keyframe k = new Keyframe(time);
            k.speed = speed;
            foreach (String part in transformations.Keys)
            {
                KeyTransformation kt = null;
                transformations.TryGetValue(part, out kt);
                k.transformations.Add(part, kt.Copy());
            }
            return k;
        }

        public void Paste(Keyframe k)
        {
            speed = k.speed;
            transformations.Clear();
            foreach (String part in k.transformations.Keys)
            {
                KeyTransformation kt = null;
                transformations.TryGetValue(part, out kt);
                transformations.Add(part, kt.Copy());
            }
        }
    }
}
